Implement retry for `ServerVersion.AutoDetect` in `AddMySqlDbContext` for Pomelo

bgrainger opened the pull request:
Fixes #1866.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2386)

bgrainger committed:
Implement retry for ServerVersion.AutoDetect. Fixes #1866

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -75,11 +76,36 @@ void ConfigureDbContext(IServiceProvider serviceProvider, DbContextOptionsBuilde
                 ConnectionStringValidation.ValidateConnectionString(settings.ConnectionString, connectionName, DefaultConfigSectionName, $"{DefaultConfigSectionName}:{typeof(TContext).Name}", isEfDesignTime: EF.IsDesignTime);
-                serverVersion = ServerVersion.AutoDetect(connectionString);
+
+                // Implement a simple backoff and retry loop based on MySqlRetryingExecutionStrategy (which has not been dependency-injected yet).
```
Can we use Polly instead? See https://github.com/dotnet/aspire/issues/2077 for some links on how to use the new version.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -75,11 +76,36 @@ void ConfigureDbContext(IServiceProvider serviceProvider, DbContextOptionsBuilde
+                {
+                    try
+                    {
+                        serverVersion = ServerVersion.AutoDetect(connectionString);
```
Are we sure we want to "double down" on AutoDetect? @lauxjpn indicated it shouldn't be used in production:

https://github.com/dotnet/aspire/pull/1161#discussion_r1488299631

bgrainger replied:
I know that was the recommendation, but from a GitHub [code search](https://github.com/search?q=UseMySql+ServerVersion+language%3AC%23&type=code&l=C%23) it looked like most users (unsurprisingly, really) were taking "the easy way out" and using `AutoDetect`. (I.e., I saw `AutoDetect` much more in those code search results than a hardcoded version).

I'm at a little bit of a loss to really know what the "right" approach is:

1. Stop using `AutoDetect` in Aspire at all: users will have to specify the server version or it will fail at startup.
2. Strongly encourage specification of a `ServerVersion` in documentation and examples (e.g., sample app uses `builder.AddMySqlDbContext<MyDbContext>("Catalog", settings => settings.ServerVersion = "8.3.0");`) but keep this current fallback code in case it's omitted.
3. Even more strongly encourage its use by making `ServerVersion` a parameter to `.AddMySqlDbContext()`.

I guess this all boils down to: keep the fallback or not. If there is no fallback, users _must_ supply `ServerVersion` and we can debate the ergonomics of how best to accomplish that. Even if we keep the fallback, we could still discuss what affordances should be provided to make it as easy as possible to fall into the pit of success. (Adding a parameter to `AddMySqlDbContext` probably isn't right, as it makes it much more difficult to use per-environment configuration.)

My gut feeling right now is that even though using AutoDetect is strongly discouraged by Pomelo, it _is_ provided and supported, and most consumers seem to use it. So for reasons of inertia, Aspire should do the same thing and we can discuss how to encourage people not to, or at least provide good examples in the documentation. Thoughts?

bgrainger committed:
Use Polly for retries.

bgrainger committed:
Switch from Microsoft.Extensions.Resilience to Polly.Extensions.

The code was not using any of the additional features of Microsoft.Extensions.Resilience.

bgrainger committed:
Use static lambdas.

This improves performance for Polly: https://www.pollydocs.org/advanced/performance.html#use-static-lambdas.

mitchdenny commented:
@eerhardt I've triaged this into P4 but could you decide whether it needs to go in after snap?

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -63,6 +67,22 @@ public static partial class AspireEFMySqlExtensions
         builder.Services.AddDbContextPool<TContext>(ConfigureDbContext);
 
+        const string resilienceKey = "Microsoft.Extensions.Hosting.AspireEFMySqlExtensions.ServerVersion";
+        builder.Services.AddResiliencePipeline(resilienceKey, static builder =>
```
Can we follow the same pattern as #2400 here?

bgrainger replied:
Yes, will do.

eerhardt replied:
I did in https://github.com/mysql-net/aspire/pull/2. If you want to take that.

bgrainger replied:
Thanks! Was tied up in meetings all morning so haven't been able to get to this. Appreciate it.

eerhardt committed:
Remove the resilience strategy from DI

bgrainger committed:
Merge pull request #2 from eerhardt/fix-serverversion

Remove the resilience strategy from DI.

bgrainger committed:
Merge main into fix-serverversion.

bgrainger committed:
Use chained syntax.

bgrainger committed:
Revert "Use chained syntax."

This reverts commit 60fd8a9769866b3a06351ef44c5ff8a958ee1133.

bgrainger committed:
Revert "Remove the resilience strategy from DI"

This reverts commit 896ca5ed0391e61622f1d6e817b3d70c0fc20220.

bgrainger committed:
Remove unnecessary variable initialization.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -68,6 +69,22 @@ public static partial class AspireEFMySqlExtensions
+                BackoffType = DelayBackoffType.Exponential,
+                MaxRetryAttempts = 6,
+                Delay = TimeSpan.FromSeconds(1),
+                MaxDelay = TimeSpan.FromSeconds(30),
```
Why set `MaxDelay` while also setting `MaxRetryAttempts`? Isn't 1 always going to be shorter than the other?

bgrainger replied:
`MaxDelay` "Caps the calculated retry delay to a specified maximum duration". (So the delays might go 1, 2, 4, 8, 16, 30, 30, 30) instead of (...16, 32, 64, 128).

I suppose in practice, since `MaxRetryAttempts` is 6, this only caps the final retry from 32s to 30s.

This was "porting" the defaults from https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/MySqlRetryingExecutionStrategy.cs and https://github.com/dotnet/efcore/blob/main/src/EFCore/Storage/ExecutionStrategy.cs#L25-L37 for lack of any better defaults to arbitrarily choose; I'm not committed to these values and am happy to remove this line if it's more confusing than anything else.

eerhardt reviewed (approved):
LGTM. Thanks for the contribution!

lauxjpn commented:
Be aware that if you use `ServerVersion.AutoDetect`, then generate migrations on your development machine, and your local and production MySQL database servers differ even slightly in their version (just one patch level can be enough), then your generated migrations might not work on your production server.

Also, generating migrations without a locally available database server does fail by default when using `ServerVersion.AutoDetect`.
